File: InvoiceManagementSystem/IMS.Infrastructure/Services/Product/ItemPriceService.cs
using IMS.Application.DTOs.Product;
using IMS.Application.Interfaces.Product;
using IMS.Application.Interfaces.Common;
using IMS.Domain.Entities.Product;
using IMS.Domain.Entities.Pricing;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace IMS.Infrastructure.Services.Product
{
    public class ItemPriceService : IItemPriceService
    {
        private readonly IRepository<ItemPrice> _repository;
        private readonly IRepository<Item> _itemRepository;
        private readonly IRepository<PriceList> _priceListRepository;

        public ItemPriceService(IRepository<ItemPrice> repository, IRepository<Item> itemRepository, IRepository<PriceList> priceListRepository)
        {
            _repository = repository;
            _itemRepository = itemRepository;
            _priceListRepository = priceListRepository;
        }

        public async Task<IEnumerable<ItemPriceDto>> GetAllAsync()
        {
            var items = await _repository.GetAllAsync(ip => ip.Item, ip => ip.PriceList);
            return items.Select(ip => MapToDto(ip, ip.Item!, ip.PriceList!)).ToList();
        }

        public async Task<ItemPriceDto?> GetByIdAsync(Guid id)
        {
            var entity = await _repository.GetByIdAsync(id, ip => ip.Item, ip => ip.PriceList);
            if (entity == null) return null;
            return MapToDto(entity, entity.Item!, entity.PriceList!);
        }

        public async Task<IEnumerable<ItemPriceDto>> GetByItemIdAsync(Guid itemId)
        {
            var items = await _repository.GetAllAsync(ip => ip.Item, ip => ip.PriceList);
            items = items.Where(ip => ip.ItemId == itemId);
            return items.Select(ip => MapToDto(ip, ip.Item!, ip.PriceList!)).ToList();
        }

        public async Task<IEnumerable<ItemPriceDto>> GetByPriceListIdAsync(Guid priceListId)
        {
            var items = await _repository.GetAllAsync(ip => ip.Item, ip => ip.PriceList);
            items = items.Where(ip => ip.PriceListId == priceListId);
            return items.Select(ip => MapToDto(ip, ip.Item!, ip.PriceList!)).ToList();
        }

        public async Task<ItemPriceDto> CreateAsync(CreateItemPriceDto dto)
        {
            var entity = new ItemPrice
            {
                Id = Guid.NewGuid(),
                ItemId = dto.ItemId,
                PriceListId = dto.PriceListId,
                Price = dto.Price,
                EffectiveFrom = dto.EffectiveFrom,
                EffectiveTo = dto.EffectiveTo,
                IsActive = true,
                IsDeleted = false
            };

            await _repository.AddAsync(entity);
            await _repository.SaveChangesAsync();

            var item = await _itemRepository.GetByIdAsync(entity.ItemId);
            var priceList = await _priceListRepository.GetByIdAsync(entity.PriceListId);
            return MapToDto(entity, item!, priceList!);
        }

        public async Task<ItemPriceDto?> UpdateAsync(Guid id, CreateItemPriceDto dto)
        {
            var entity = await _repository.GetByIdAsync(id);
            if (entity == null) return null;
            if (entity.IsDeleted || !entity.IsActive) return null;

            entity.ItemId = dto.ItemId;
            entity.PriceListId = dto.PriceListId;
            entity.Price = dto.Price;
            entity.EffectiveFrom = dto.EffectiveFrom;
            entity.EffectiveTo = dto.EffectiveTo;

            _repository.Update(entity);
            await _repository.SaveChangesAsync();

            var item = await _itemRepository.GetByIdAsync(entity.ItemId);
            var priceList = await _priceListRepository.GetByIdAsync(entity.PriceListId);
            return MapToDto(entity, item!, priceList!);
        }

        public async Task<bool> DeleteAsync(Guid id)
        {
            var entity = await _repository.GetByIdAsync(id);
            if (entity == null) return false;
            if (entity.IsDeleted) return false;

            _repository.Delete(entity);
            await _repository.SaveChangesAsync();
            return true;
        }

        private ItemPriceDto MapToDto(ItemPrice ip, Item i, PriceList p)
        {
            return new ItemPriceDto
            {
                Id = ip.Id,
                Price = ip.Price,
                EffectiveFrom = ip.EffectiveFrom,
                EffectiveTo = ip.EffectiveTo,
                Item = new ItemDto
                {
                    Id = i.Id,
                    Name = i.Name,
                    SKU = i.SKU
                },
                PriceList = new IMS.Application.DTOs.Pricing.PriceListDto
                {
                    Id = p.Id,
                    Name = p.Name,
                    IsDefault = p.IsDefault
                }
            };
        }
    }
}

File: InvoiceManagementSystem/ims.ClientApp/src/app/product/item-price/item-price.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { Item } from '../../product/items/item.service';

export interface PriceList {
    id: string;
    name: string;
    isDefault: boolean;
}

export interface ItemPrice {
    id: string;
    item?: Item | null;
    priceList?: PriceList | null;
    price: number;
    effectiveFrom: string;
    effectiveTo?: string;
}

export interface CreateItemPriceDto {
    itemId: string;
    priceListId: string;
    price: number;
    effectiveFrom: string;
    effectiveTo?: string;
}

@Injectable({
    providedIn: 'root'
})
export class ItemPriceService {
    private apiUrl = 'https://localhost:7276/api/itemprice';
    private priceListUrl = 'https://localhost:7276/api/pricelist';
    private itemUrl = 'https://localhost:7276/api/item';

    constructor(private http: HttpClient) { }

    getAll(): Observable<ItemPrice[]> {
        return this.http.get<any[]>(this.apiUrl).pipe(
            map(data => data.map(ip => ({
                id: ip.id || ip.Id,
                price: ip.price || ip.Price,
                effectiveFrom: ip.effectiveFrom || ip.EffectiveFrom,
                effectiveTo: ip.effectiveTo || ip.EffectiveTo,
                item: ip.item || ip.Item ? {
                    id: (ip.item || ip.Item).id || (ip.item || ip.Item).Id,
                    name: (ip.item || ip.Item).name || (ip.item || ip.Item).Name,
                    sku: (ip.item || ip.Item).sku || (ip.item || ip.Item).SKU,
                    unitOfMeasureId: (ip.item || ip.Item).unitOfMeasureId || (ip.item || ip.Item).UnitOfMeasureId
                } : null,
                priceList: ip.priceList || ip.PriceList ? {
                    id: (ip.priceList || ip.PriceList).id || (ip.priceList || ip.PriceList).Id,
                    name: (ip.priceList || ip.PriceList).name || (ip.priceList || ip.PriceList).Name,
                    isDefault: (ip.priceList || ip.PriceList).isDefault || (ip.priceList || ip.PriceList).IsDefault
                } : null
            } as unknown as ItemPrice)))
        );
    }

    getById(id: string): Observable<ItemPrice> {
        return this.http.get<ItemPrice>(`${this.apiUrl}/${id}`);
    }

    create(dto: CreateItemPriceDto): Observable<ItemPrice> {
        return this.http.post<ItemPrice>(this.apiUrl, dto);
    }

    update(id: string, dto: CreateItemPriceDto): Observable<ItemPrice> {
        return this.http.put<ItemPrice>(`${this.apiUrl}/${id}`, dto);
    }

    delete(id: string): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`);
    }

    getItems(): Observable<Item[]> {
        return this.http.get<Item[]>(this.itemUrl);
    }

    getItemsWithPricesForPriceList(priceListId: string): Observable<any[]> {
        return this.http.get<any[]>(`${this.apiUrl}/pricelist/${priceListId}/items`);
    }

    getPriceLists(): Observable<PriceList[]> {
        return this.http.get<PriceList[]>(this.priceListUrl);
    }
}
